src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -366,36 +361,54 @@ public async Task SaveChangesAsync(IEnumerable<UnspentOutputs> unspentOutputs, I
 
                 foreach (UnspentOutputs unspent in unspentOutputs)
                 {
-                    if (this.unspents.TryGetValue(unspent.TransactionId, out CacheItem cacheItem))
+                    if (!this.unspents.TryGetValue(unspent.TransactionId, out CacheItem cacheItem))
                     {
-                        // We'll need to restore the original outputs, so we clone it
-                        // and save it in rewind data.
-                        UnspentOutputs clone = unspent.Clone();
+                        // This can happen very rarely in case we fetch items form
+                        // disk and immedietly after the Evict method is called and removes the cached items.
 
-                        if (cacheItem.UnspentOutputs != null)
-                        {
-                            clone.Outputs = cacheItem.UnspentOutputs.Outputs.ToArray();
-                            rewindData.OutputsToRestore.Add(clone);
-                        }
-                        else
-                        {
-                            rewindData.TransactionsToRemove.Add(unspent.TransactionId);
-                        }
+                        this.logger.LogTrace("Outputs of transaction ID '{0}' are not found in cache, creating them.", unspent.TransactionId);
 
+                        FetchCoinsResponse result = await this.inner.FetchCoinsAsync(new[] {unspent.TransactionId}).ConfigureAwait(false);
+
+                        UnspentOutputs unspentOutput = result.UnspentOutputs[0];
+
+                        cacheItem = new CacheItem();
+                        cacheItem.ExistInInner = unspentOutput != null;
+                        cacheItem.IsDirty = false;
+
+                        cacheItem.UnspentOutputs = unspentOutput?.Clone();
+
+                        this.unspents.TryAdd(unspent.TransactionId, cacheItem);
```

Comment:
Can we perhaps rename `unspents` to be more specific? I.e. I'm not always sure which unspent set we are talking about. Perhaps `cachedUTXOset? or `cachedUnspentItems`?